research/02-repl.tex

```diff
@@ -39,94 +39,117 @@ \subsection{Origin of REPLs}
 \label{ssec:repl-advantages}
-REPLs offer several advantages that make them an often requested feature.
-Because of their nature, REPLs provide the ability to program interactively.
-Programming interactively has multiple applications.
-
-When creating software solutions for an as of yet not well understood domain,
-it is often not clear which data structures and algorithms are required. In such
-cases, REPLs offer the ability to interactively develop and debug software
-without having to apply the (oftentimes much slower) edit-compile-run-debug
-development style. This kind of programming is called exploratory
-programming~\cite{Fritzson86}. Related to this kind of exploration, REPLs also
-provide a means for rapid prototyping and bottom up programming~\cite{Graham93}.
+
+REPLs provide the ability to program interactively. Programming interactively
+has multiple advantages.
+
+When creating software solutions for an as of yet not well understood domain, it
+is often not clear which data structures and algorithms are required. In such
+cases, interactively developing and debugging software is an advantage over the
+(oftentimes much slower) edit-compile-run-debug development style. This kind of
+programming is called exploratory programming~\cite{Fritzson86}. Related to this
+kind of exploration, programming interactively also provides a means for rapid
+prototyping and bottom up programming~\cite{Graham93}.
 
 The explorative and interactive features of a REPL also make it an excellent
 tool for programmers to learn a new programming language. REPLs are also
 combined with what is called literate programming to offer notebooks or language
 playgrounds, as discussed in \cref{sec:literate-programming}.
 
+\subsection{Execution model}
+\label{ssec:execution-model}
+
+Every programming language defines an execution model, which specifies how
+programs written in that language are executed. Amongst others, it specifies
+what an indivisable amount of work is (a \emph{compilation unit}) and in what
+order these units are executed. The implementation of an execution model is a
+compiler and/or an interpreter, often accompanied with a runtime system to
+provide features such as garbage collection.
+
+The execution model as implemented by a REPL differs only slightly to that of an
+interpreter or compiler. A difference might be as to what is considered an
+indivisable unit of work: a REPL might accept singular expressions that a
+regular interpreter or compiler might not. Python's execution model, for
+example, states the following: \textit{``A block is a piece of Python program
+text that is executed as a unit. The following are blocks: a module, a function
+body, and a class definition. Each command typed interactively is a
+block''}~\cite{PythonExecutionModel}. This implies that the interactive
+interpreter acceps smaller blocks than the non-interactive interpreter.
+
+A compiler or an interpreter reads the files that make up a program's source
```

Comment:
I like the beginning of your suggestion. The end however is already too specifically geared towards Spoofax and is more fit for the problem analysis section.

I would also like to keep the part about REPLs having to keep everything in memory, but we can discuss and improve this part tomorrow.